feat: introduced cloud states

kittaakos opened the pull request:
### Motivation
<!-- Why this pull request? -->

Please reference the internal slides regarding the expected behavior.


https://user-images.githubusercontent.com/1405703/221146190-6c99551a-3e6f-4934-a9f7-321b1b124f6f.mp4


https://user-images.githubusercontent.com/1405703/221146231-1529b638-f7c4-4057-ac57-9864167f0549.mp4


https://user-images.githubusercontent.com/1405703/221161339-1620473f-f8e9-4c7d-b50b-ac9fe516fef3.mp4


https://user-images.githubusercontent.com/1405703/221516692-36a0eb8b-ccc3-4a33-a0c6-7e37930e56f9.mp4



### Change description
<!-- What does your code do? -->

 - New UX for the sketchbook,
 - Show `[Cloud]` prefix in the window title,
 - Pull/push sketch from the editor toolbar, and
 - Create a cloud copy of the local sketch.

### Other information
<!-- Any additional information that could help the review process -->

Closes #1879
Closes #1876
Closes #1899
Closes #1878

~This PR includes https://github.com/arduino/arduino-ide/pull/1910/commits/3f5eddb033c3e133e753606b8de9a20828fcd58b from https://github.com/arduino/arduino-ide/pull/1910.~

### Reviewer checklist

* [ ] PR addresses a single concern.
* [ ] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-ide/pulls) before creating one)
* [ ] PR title and description are properly filled.
* [ ] Docs have been added / updated (for bug fixes / features)

kittaakos commented on arduino-ide-extension/src/browser/contributions/create-cloud-copy.ts:
```diff
@@ -0,0 +1,117 @@
+}
+function isCreateCloudCopyParams(arg: unknown): arg is CreateCloudCopyParams {
+  return (
+    (<CreateCloudCopyParams>arg).model !== undefined &&
```
guard against falsy

kittaakos replied:
https://github.com/arduino/arduino-ide/pull/1901/commits/7ac2c57e1cef7f4292d96a3bc2ac586ff14f0003

kittaakos commented on arduino-ide-extension/src/browser/create/create-features.ts:
```diff
@@ -64,14 +76,55 @@ export class CreateFeatures implements FrontendApplicationContribution {
+    return this._enabled;
+  }
+
+  get cloudSketchStates(): {
```
unused. remove

kittaakos replied:
https://github.com/arduino/arduino-ide/pull/1901/commits/e36127aed519f70bbcfaf26d8197c00559d2bf79

kittaakos commented on arduino-ide-extension/src/browser/theia/core/window-title-updater.ts:
```diff
@@ -22,12 +32,22 @@ export class WindowTitleUpdater extends TheiaWindowTitleUpdater {
 
   private readonly applicationName =
     FrontendApplicationConfigProvider.get().applicationName;
+  private readonly toDispose = new DisposableCollection();
```
dispose on FE stop

kittaakos commented on arduino-ide-extension/src/browser/theia/core/window-title-updater.ts:
```diff
@@ -1,18 +1,28 @@
+import { nls } from '@theia/core';
```
expand it

kittaakos replied:
https://github.com/arduino/arduino-ide/pull/1901/commits/6471e1dc0813a1309da9b6b69d9bff378e05a0b3

AlbyIanna commented:
I've stumbled in an edge case that resulted in having a broken cloud user icon and an empty cloud sketchbook. Steps to reproduce:
- make sure your internet connection is ON
- sign in to cloud from the cloud sketchbook
- make sure you have pulled at least one sketch (e.g.: `sketch_A`)
- turn OFF your internet connection
- open a sketch from the cloud sketchbook that you already had pulled (e.g. `sketch_A`)
- a new sketch window opens showing the cloud sketch
- 🐛 in the left sidebar, the user image is broken (I guess we should have a default image to show when we can't show the actual user image)
- 🐛 the cloud sketchbook is empty


<details>
  <summary>Here's a recording</summary>

https://user-images.githubusercontent.com/6939054/222192264-fd426c7f-4645-4584-9ce1-65195a1b54fa.mov

</details>

I'm not sure what the expected behaviour should be here. If possible I guess we should load the cloud sketchbook in the same state as it is in the current window, but I'm not sure we can do that.

@kittaakos thoughts?

AlbyIanna commented:
Another thing I stumbled into, but I'm not sure it's an unexpected behaviour or if I am missing something about the expected behaviour:

In the following 2 screen recordings, I'm pushing sketches from my local sketchbook to the cloud sketchbook.

The first one results in a different behaviour from the the second one, asking for user's confirmation before pushing the sketch, and after that opening the pushed sketch in a new window.

<details>
<summary>screen recording 1</summary>

https://user-images.githubusercontent.com/6939054/222206186-796bacbf-b049-4048-bd70-0ca86ec47f93.mov

</details>

The second one doesn't ask for confirmation and doesn't open the sketch in a new window:

<details>
<summary>screen recording 2</summary>

https://user-images.githubusercontent.com/6939054/222206160-ad270aad-f719-4a7c-b2d8-f9b3ad866ebc.mov

</details>

I'm not sure about the confirmation message, but in the second recording I think it's expected to open the pushed sketch in a new window.

AlbyIanna commented on arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-widget.tsx:
```diff
@@ -134,26 +133,37 @@ export class SketchbookTreeWidget extends FileTreeWidget {
     if (SketchbookTree.SketchDirNode.is(node) && node.commands) {
       return Array.from(new Set(node.commands)).map((command) =>
-        this.renderInlineCommand(command.id, node)
+        this.renderInlineCommand(
+          Array.isArray(command)
+            ? command
+            : typeof command === 'string'
+            ? command
+            : command.id,
+          node
+        )
```
I think we can handle this condition inside `renderInlineCommand` and just pass the command like this:

```suggestion
        this.renderInlineCommand(command, node)
```

kittaakos replied:
It's [called from the subclass with the command ID](https://github.com/arduino/arduino-ide/blob/e36127aed519f70bbcfaf26d8197c00559d2bf79/arduino-ide-extension/src/browser/widgets/cloud-sketchbook/cloud-sketchbook-tree-widget.tsx#L124), I did not want to change unrelated code, but I can do if you think it has value.

AlbyIanna commented on arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-widget.tsx:
```diff
@@ -134,26 +133,37 @@ export class SketchbookTreeWidget extends FileTreeWidget {
   ): React.ReactNode {
-    const command = this.commandRegistry.getCommand(commandId);
-    const icon = command?.iconClass;
+    const commandId = Array.isArray(command) ? command[0] : command;
```
Follow-up comment of the one I made in `renderInlineCommands`:
if you change the type of the `command` argument to accept a `Command | string | [command: string, label: string]`, you can move the check I mentioned here, like this:
```suggestion
    const commandId = Command.is(command)
      ? command.id
      : Array.isArray(command)
      ? command[0]
      : command;
```

AlbyIanna reviewed (commented):
Thank you @kittaakos for this! Overall quality of the code is outstanding, as always. 

I just left some of comments that in my opinion could improve the readability. Let me know what you think about them. I didn't request changes because they're not blocking.

Apart from those, I also left a couple of comments about unexpected behaviour.

kittaakos commented on arduino-ide-extension/src/browser/theia/messages/notifications-manager.ts:
```diff
@@ -46,11 +45,4 @@ export class NotificationManager extends TheiaNotificationManager {
     }
     return Math.min((update.work.done / update.work.total) * 100, 100);
   }
-
-  /**
-   * For `public` visibility.
-   */
-  override getMessageId(message: Message): string {
-    return super.getMessageId(message);
-  }
```
> anyway private method

Because it has `protected` visibility and accessing `protected` props this way, neither requires casting to `any` nor relaxing the `@typescript-eslint/dot-notation` ES-Lint rule with `{ "allowPrivateClassPropertyAccess": true }`.

kittaakos commented:
> * 🐛 in the left sidebar, the user image is broken (I guess we should have a default image to show when we can't show the actual user image)

I added a patch and the default account codicon is visible when loading the avatar has failed.

kittaakos commented:
> I'm not sure about the confirmation message, but in the second recording I think it's expected to open the pushed sketch in a new window.

Thanks! I could reproduce it when creating a cloud copy without renaming the local sketch. The problem was in `ncp`, so I switched to a new lib and added a few tests.

kittaakos commented:
@91volt, can you please help with the review? Thanks!

kittaakos commented:
Thanks for the review!

> * ❌ this is still happening
>   > 🐛 the cloud sketchbook is empty


> Clicking on the "Refresh" button will have the cloud sketchbook loaded, but the user image will still remain unavailable.

I expected it would be OK. What is your expectation?

91volt commented:
>✅ this is fixed (now it always asks for confirmation before pushing a sketch)

I suggest removing the confirmation prompt when copying a sketch to the cloud, as it may cause confusion for the user since there is no existing sketch being overwritten. Additionally, both options currently result in the same outcome.

About this:
>🐛 the cloud sketchbook is empty

>Clicking on the "Refresh" button will have the cloud sketchbook loaded.

>I expected it would be OK. What is your expectation?

Ideally, this glitch wouldn't exist at all, or the sketchbook would be refreshed automatically in such cases. If fixing it now is time-consuming, we could consider publishing a separate fix for this bug since it is not critical for the general purpose of this feature in my opinion.

AlbyIanna commented:
>> Clicking on the "Refresh" button will have the cloud sketchbook loaded, but the user image will still remain unavailable.

> I expected it would be OK. What is your expectation?

About this, I agree with the last comment from @91volt ⬇️ 
> Ideally, this glitch wouldn't exist at all, or the sketchbook would be refreshed automatically in such cases. If fixing it now is time-consuming, we could consider publishing a separate fix for this bug since it is not critical for the general purpose of this feature in my opinion.

AlbyIanna commented:
> I suggest removing the confirmation prompt when copying a sketch to the cloud, as it may cause confusion for the user since there is no existing sketch being overwritten. Additionally, both options currently result in the same outcome.

I too was thinking it would be a good idea to remove it. In the end, it's not possible to push a local sketch if it already exists in the cloud, therefore it doesn't make much sense to warn the user saying that "Pushing this Sketch will overwrite its Cloud version."

kittaakos commented:
> * ❌ this is still happening
>   > 🐛 the cloud sketchbook is empty

Thanks! I fixed it

https://user-images.githubusercontent.com/1405703/224361090-ca63a11f-c1a0-4463-a339-6d62bc0121a6.mp4

> as it may cause confusion for the user since there is no existing sketch being overwritten.

Technically, there is. There is no such thing as _create a sketch with these files_ endpoint. IDE2 has to create a default one and push the local files, so it's not an atomic operation. I tame IDE2 to ignore the `arduino.cloud.push.warn` preference for this use case.

AlbyIanna reviewed (approved):
🟢 Green light for me!

Thank you Akos!

per1234 commented on arduino-ide-extension/src/browser/theia/core/connection-status-service.ts:
```diff
@@ -1,106 +1,324 @@
+    if (!this.isOnline.online) {
+      const text = nls.localize(
+        'arduino/connectionStatus/connectionLost',
+        "Connection lost. Cloud Sketches actions and updates won't be available."
```
```suggestion
        "Connection lost. Cloud sketch actions and updates won't be available."
```

- "Sketches" -> "sketches": The decision was made that "sketch" is not a proper noun and thus should not be capitalized in sentence case text.
  reference: the "Arduino Editorial Guidelines" document in the ATL Confluence space
- "sketches" -> "sketch": Although the other isn't necessarily wrong, I think my suggestion is slightly cleaner grammatically.

kittaakos replied:
Thank you!

I decided to make the changes manually instead of committing the proposed change. If I commit this, I must still manually regenerate the `en.json` file with `yarn i18n:generate`.


https://user-images.githubusercontent.com/1405703/225549692-3d7e2871-70a5-45e4-8635-8aeb6ed45bfa.mp4

kittaakos commented:
Thanks to all of you for the excellent reviews.
